cpp/src/arrow/ipc/options.h

```diff
@@ -44,8 +59,31 @@ struct ARROW_EXPORT IpcOptions {
   /// consisting of a 4-byte prefix instead of 8 byte
   bool write_legacy_ipc_format = false;
 
+  /// \brief The memory pool to use for allocations made during IPC writing
+  MemoryPool* memory_pool = default_memory_pool();
+
+  /// \brief EXPERIMENTAL: Codec to use for compressing and decompressing
+  /// record batch body buffers. This is not part of the Arrow IPC protocol and
+  /// only for internal use (e.g. Feather files)
+  Compression::type compression = Compression::UNCOMPRESSED;
+  int compression_level = Compression::kUseDefaultCompressionLevel;
+
   static IpcOptions Defaults();
 };
 
+struct ARROW_EXPORT IpcReadOptions {
+  // The maximum permitted schema nesting depth.
+  int max_recursion_depth = kMaxNestingDepth;
+
+  /// \brief The memory pool to use for allocations made during IPC writing
+  MemoryPool* memory_pool = default_memory_pool();
+
+  /// \brief EXPERIMENTAL: Top-level schema fields to include when
+  /// deserializing RecordBatch. Null means to return all deserialized fields
+  std::shared_ptr<std::unordered_set<int>> included_fields;
```

Comment:
I wanted to amortize the creation of the set. 

For example, suppose you select 500 out of 1000 fields during deserialization and have 1000 batches. If you use `std::vector` then either you create the set 1000 times or you use `std::find` with `std::vector` and live with linear time lookups. Neither is a very good option.

Aside, I'm wondering if we should create a simple fixed-size-hash-set implementation that we can use in this project to avoid including `unordered_set`?

Reply:
Well, if there was some kind of IPCReader that you could instantiate and that would keep the desired data structure under the hood, we wouldn't have this problem :-)

In this case and if you want fast lookups, the best is probably a `std::vector<bool>`...

Reply:
Alright, I'll refactor to use `std::vector<bool>`

Reply:
I meant use `vector<bool>` internally if you want fast lookups. I think the user-exposed `IpcReadOptions` would be more ergonomic with a `vector<int>`.

Reply:
Hm, okay. This isn't that trivial to do because we have single-use `ReadRecordBatch` functions as well as the stream reader classes that read multiple record batches. I'll see what I can do

Reply:
FWIW: I think a `vector<bool>` is perfectly acceptable as a user-exposed `set<int>`, though doing so warrants some helpers like `vector<bool> FromIntVector(const vector<int>&)`, `string PrintSetIndices(const vector<bool>&)`

Reply:
I went ahead and implemented the `std::vector<int>` version for now